varargs for einsum

t-vi opened the pull request:
Implemented via a wrapper, thank you Richard for the suggestion!

Fixes: #9929

t-vi committed:
varargs for einsum

implemented via a wrapper

Fixes: #9929

fmassa commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -530,7 +530,7 @@
   variants: function
 
-- func: einsum(std::string equation, TensorList tensors) -> Tensor
+- func: _einsum(std::string equation, TensorList tensors) -> Tensor
```
can't we rename the Python wrapper in here, instead of renaming the function in ATen?

t-vi replied:
I don't think the Python function should be called anything but einsum.
But we could keep the ATen name and use  torch._C.VariableFunctions.einsum from the Python wrapper and let the wrapper be torch.einsum in Python.

fmassa replied:
I'm not sure if the functionality I was thinking is available at this level, but it was possible at some point to expose a different python name than the C name of the function, similar to [what is present here](https://github.com/pytorch/pytorch/blob/master/tools/autograd/deprecated.yaml#L7-L8).
So I was thinking about keeping the ATen name `einsum`, and having this Python wrapper be called `_einsum`, so that the rest of the PR doesn't change.

t-vi committed:
Keep einsum name in ATen

Thanks to fmassa for the suggestion!
if the method to fix it is not as pretty, it is my own fault, though.

fmassa commented on torch/functional.py:
```diff
@@ -140,6 +141,80 @@ def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+"""
+    if len(operands) == 1 and isinstance(operands[0], (list, tuple)):
+        # the old interface of passing the operands as one list argument
+        operands = operands[0]
```
question: do we want to deprecate this behavior? just wondering

t-vi replied:
I don't know and will adjust to your preference. The C++ side will still have it.

zou3519 commented on torch/functional.py:
```diff
@@ -140,6 +141,80 @@ def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
     return P, L, U
 
 
+def einsum(equation, *operands):
```
Hmm it seems unfortunate that we have to move einsum into torch.nn.functional.einsum to use the varargs feature.

t-vi commented:
We don't, torch.functional.einsum (no nn) is imported as torch.einsum.

zou3519 commented:
Oh I see. Thanks for the clarification @t-vi !

t-vi commented:
I think the CI failure isn't this patch. Should we deprecate having a list as second argument?

zou3519 commented:
Let's keep it like this for now and think about if the deprecation should happen later.

soumith committed:
Merge branch 'master' into einsum_varargs
